### PR TITLE
Deployment preparation: Gnosis Root Gauge Factory

### PR DESCRIPTION
**Pull request by EndymionJkb**
# Deployment preparation: 

Deployment of the Gnosis Root Gauge Factory. build-info was generated from commit [cc0edea7fbec4aa2db75abf618fb47c0d364b2ff](https://github.com/balancer-labs/balancer-v2-monorepo/commit/cc0edea7fbec4aa2db75abf618fb47c0d364b2ff)

## Deployment preparation tasks checklist:

- [X] Deployments `README` is updated with the new task name and link
- [X] Task `README` has a description and links to the artifacts and expected output folders 
- [X] `build-info` is updated with artifacts compiled at the specified commit
- [X] Artifacts are generated from `build-info`
- [ ] New task has a working fork test 

## Code checklist:

- [X] The diff is legible and has no extraneous changes 
- [N/A] Complex code has been commented in tests and scripts
- [X] Fork test block number and test cases are appropriate for this deployment
- [X] The base branch is either `master`, or there's a description of how to merge

### CODE REVIEW

**Comment by jubeira** [comment]
I've checked the code @https://github.com/balancer-labs/balancer-v2-monorepo/pull/2251/commits/342507d24fffb602d523cdd306798f284ddecc05 and the build info committed here matches the one generated by https://github.com/balancer-labs/balancer-v2-monorepo/commit/cc0edea7fbec4aa2db75abf618fb47c0d364b2ff as stated in the description.

I've done a first pass and added a handful of comments about the fork test; otherwise it's looking good!

File: pkg/deployments/tasks/20230217-gnosis-root-gauge-factory/test/task.fork.ts [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+      },
+      { value: VAULT_BOUNTY }
+    );
+
+    await bal80weth20Pool.connect(veBALHolder).approve(veBAL.address, MAX_UINT256);
+    const currentTime = await currentTimestamp();
+    await veBAL
+      .connect(veBALHolder)
+      .create_lock(await bal80weth20Pool.balanceOf(veBALHolder.address), currentTime.add(MONTH * 12));
+  });
```
Comment (nventuro) [inline pkg/deployments/tasks/20230217-gnosis-root-gauge-factory/test/task.fork.ts:118]: This is all a bit convoluted - you can just impersonate any veBAL holder and use their address, and you'll skip having to join, lock, etc.

File: pkg/deployments/tasks/20230217-gnosis-root-gauge-factory/test/task.fork.ts [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+    BAL = await BALTokenAdmin.getBalancerToken();
+
+    const gaugeControllerTask = new Task('20220325-gauge-controller', TaskMode.READ_ONLY, getForkedNetwork(hre));
+    gaugeController = await gaugeControllerTask.deployedInstance('GaugeController');
+  });
+
+  before('update balancer token admin rate', async () => {
+    await advanceTime(WEEK * 5);
+    await BALTokenAdmin.update_mining_parameters();
+  });
```
Comment (nventuro) [inline pkg/deployments/tasks/20230217-gnosis-root-gauge-factory/test/task.fork.ts:90]: This is also a bit odd, why move forward into the future so much? And why manually update the minim params? It'll be done automatically

File: pkg/deployments/tasks/20230217-gnosis-root-gauge-factory/test/task.fork.ts [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+        await authorizerAdaptor.getActionId(gaugeController.interface.getSighash('add_type(string,uint256)')),
+        admin.address
+      );
+
+    await authorizer
+      .connect(govMultisig)
+      .grantRole(
+        await authorizerAdaptor.getActionId(gaugeController.interface.getSighash('add_gauge(address,int128)')),
+        admin.address
+      );
```
Comment (nventuro) [inline pkg/deployments/tasks/20230217-gnosis-root-gauge-factory/test/task.fork.ts:162]: There's no support for gnosis in the adder, and it currently doesn't work. We've also tested the adder throughly, so there's no point in doing that again here. Just add it manually by calling `gc.add_gauge`.

File: pkg/deployments/tasks/20230217-gnosis-root-gauge-factory/test/task.fork.ts [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+      await gaugeController['gauge_relative_weight(address,uint256)'](gauge.address, await currentWeekTimestamp())
+    ).to.be.gt(weightCap);
+    expect(await gauge.getCappedRelativeWeight(await currentTimestamp())).to.equal(weightCap);
+  });
+
+  it('mint & bridge tokens', async () => {
+    // The gauge has votes for this week, and it will mint the first batch of tokens. We store the current gauge
+    // relative weight, as it will change as time goes by due to vote decay.
+    const firstMintWeekTimestamp = await currentWeekTimestamp();
+    //const gaugeRelativeWeight = await gaugeController['gauge_relative_weight(address)'](gauge.address);
```
Comment (nventuro) [inline pkg/deployments/tasks/20230217-gnosis-root-gauge-factory/test/task.fork.ts:216]: Why is this commented?

File: pkg/deployments/tasks/20230217-gnosis-root-gauge-factory/test/task.fork.ts [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+    // And the gauge then deposits those in the predicate via the bridge mechanism
+    const bridgeInterface = new ethers.utils.Interface([
+      'event TokensBridgingInitiated(address indexed token, address indexed sender, uint256 value, bytes32 indexed messageId)',
+    ]);
+
+    expectEvent.inIndirectReceipt(await mintTx.wait(), bridgeInterface, 'TokensBridgingInitiated', {
+      token: BAL,
+      sender: gauge.address,
+      value: actualEmissions,
+    });
```
Comment (nventuro) [inline pkg/deployments/tasks/20230217-gnosis-root-gauge-factory/test/task.fork.ts:263]: Given that it's hard to do local tests for bridges, we're a bit more strict in the procedure for merging fork test PRs that interact with them.

Please manually initiate the bridge process from an EOA, calling the same function our code does with the same arguments (`_gnosisBridge.relayTokens`), and submit the explorer link here, showing that it emits the same event we test for here with the same arguments. Ideally we'd also see e.g. the recipient.

Then, once the bridging is complete, also paste here the explorer link to the gnosis transaction where the tokens landed, so we can verify that this call indeed works.

File: pkg/deployments/tasks/20230217-gnosis-root-gauge-factory/test/task.fork.ts (hunk 1 above)
Comment (EndymionJkb) [reply pkg/deployments/tasks/20230217-gnosis-root-gauge-factory/test/task.fork.ts:118]: It was not trivial to find one with enough veBAL to make the voting work, and it also constrains it to be at a certain block. This way we don't have to worry about that, and can do it any time.

File: pkg/deployments/tasks/20230217-gnosis-root-gauge-factory/test/task.fork.ts (hunk 2 above)
Comment (EndymionJkb) [reply pkg/deployments/tasks/20230217-gnosis-root-gauge-factory/test/task.fork.ts:90]: It's to make it not cross an epoch, which throws off the expected emissions and causes the "multiple week" minting test to fail. If I don't do this, it's off by 6%.

File: pkg/deployments/tasks/20230217-gnosis-root-gauge-factory/test/task.fork.ts (hunk 4 above)
Comment (EndymionJkb) [reply pkg/deployments/tasks/20230217-gnosis-root-gauge-factory/test/task.fork.ts:216]: I believe it was leftover from Tom's old PR, which I had to update looking at more recent ones.

File: pkg/deployments/tasks/20230217-gnosis-root-gauge-factory/test/task.fork.ts (hunk 5 above)
Comment (EndymionJkb) [reply pkg/deployments/tasks/20230217-gnosis-root-gauge-factory/test/task.fork.ts:263]: Call to relayTokens, showing the event:
https://etherscan.io/tx/0x6a0dcbf72db757f83bf1c9b42e5f940c31e3240479614635fcbe5a5f72091692

Gnosis chain receipt:
https://gnosisscan.io/tx/0x983bede2006468b5fbc67204068c2b1c9492be65530ce7f34e82841703714294

**Review by nventuro — APPROVED** [review]
I pushed a commit extending the fork test, so that we also test the recipient argument.

**Pull request merged**